### PR TITLE
[Feature:TAGrading] Download PDF Annotations

### PR DESCRIPTION
**Pull request by jchulton**
### Please check if the PR fulfills these requirements:

* [ ] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
You cannot download a pdf with annotations

### What is the new behavior?
When an instructor saves a pdf, a pdf with the annotations drawn on it will be saved to {course_path}/annotated_pdfs/{gradeable_id}/{student_id}/{version_num}/{name of pdf}_{grader_id}.pdf. A student can download this pdf when they view their grade.

Steps to replicate:
1.Create a gradeable which takes student input.
2. Submit a pdf as a student.
3. Annotate the pdf and grade the assignment as an instructor.
4. Make grades available
5. As a student, view grade and download annotated pdf.

### CODE REVIEW

**Comment by matgar1997** [comment]
Is there anything extra involved annotating. after drawing, adding some text and pressing save, the text in the box said not saved

**Comment by matgar1997** [comment]
It might be an issue with the steps to reproduce but every time I press save after annotating it gives me an undefined error popup

File: site/app/templates/grading/electronic/PDFAnnotationEmbedded.twig [hunk 1]
```diff
@@ -4,6 +4,8 @@
 {% for script in localjs %}
     <script src="{{ script }}"></script>
 {% endfor %}
+<script src="https://code.jquery.com/jquery-1.12.4.min.js"></script>
```
Comment (MasterOdin) [inline site/app/templates/grading/electronic/PDFAnnotationEmbedded.twig:7]: why is jQuery 1.12.4 here? Does jspdf work on newer versions of jQuery?

Comment (jchulton) [reply site/app/templates/grading/electronic/PDFAnnotationEmbedded.twig]: Yes it does work with newer versions. Removed the line

File: site/app/templates/grading/electronic/PDFAnnotationEmbedded.twig [hunk 2]
```diff
@@ -4,6 +4,8 @@
 {% for script in localjs %}
     <script src="{{ script }}"></script>
 {% endfor %}
+<script src="https://code.jquery.com/jquery-1.12.4.min.js"></script>  
+<script src="https://cdnjs.cloudflare.com/ajax/libs/jspdf/1.3.2/jspdf.min.js"></script>
```
Comment (jchulton) [inline site/app/templates/grading/electronic/PDFAnnotationEmbedded.twig:7]: This is probably not how we want to load this library. We could download it locally instead into the libraries folder. Where exactly should the library go?

Comment (MasterOdin) [reply site/app/templates/grading/electronic/PDFAnnotationEmbedded.twig:7]: Same as any other dependencies installeable by npm. Add it to package.json, modify .setup/install_submitty/install_site_*.sh

File: site/app/views/PDFView.php [hunk 3]
```diff
@@ -38,7 +38,9 @@ public function showPDFEmbedded($params) {
         $localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('pdfjs', 'pdf.worker.min.js'), 'vendor');
         $localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('pdf-annotate.js', 'pdf-annotate.min.js'), 'vendor');
         $localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('pdf', 'PDFAnnotateEmbedded.js'), 'js');
-
+        $localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('html2canvas', 'html2canvas.js'), 'vendor');
```
Comment (shailpatels) [inline site/app/views/PDFView.php:41]: you can use 
```php
$this->core->getOutput()->addVendorJs(FileUtils::joinPaths('html2canvas', 'html2canvas.js'));

```

We shouldn't be calling the timestampResource directly here but that can be handled when pdf annotation is re-written

File: site/public/js/pdf/PDFInitToolbar.js [hunk 4]
```diff
@@ -146,33 +147,101 @@ window.onbeforeunload = function() {
     }
 
     function saveFile(){
```
Comment (shailpatels) [inline site/public/js/pdf/PDFInitToolbar.js:149]: can you add some comments explaining whats going on in these functions.

File: site/public/js/pdf/PDFInitToolbar.js [hunk 5]
```diff
@@ -146,33 +147,101 @@ window.onbeforeunload = function() {
...
+                        let ctx = canvas.getContext('2d');
+                        let annotationLayer = pageContainer.getElementsByClassName("annotationLayer")[0];
+                        let scale = annotationLayer.scale;
+                        let children = annotationLayer.children;
+                        ctx.scale(zoom_level/100, zoom_level/100);
+                            for (let j = 0; j < children.length; j++) {
+                                let annotation = children[j];
+                                if(annotation.getAttribute("data-pdf-annotate-type") == "drawing"){
+                                    let path2 = new Path2D(annotation.getAttribute("d"));
+                                    ctx.strokeStyle = annotation.getAttribute("stroke");
```
Comment (shailpatels) [inline site/public/js/pdf/PDFInitToolbar.js:217]: It looks like the annotations are getting read from the annotation layer and then getting drawn onto the canvas which is getting converted to an image, is there a way to directly take the annotations and use pdfJS to bake the annotations back into the pdf?

Is this something we can use http://raw.githack.com/MrRio/jsPDF/master/docs/module-context2d.html ?

File: site/app/controllers/MiscController.php [hunk 6]
```diff
@@ -179,10 +180,10 @@ public function downloadCourseFile($dir, $path) {
         // security check
         $path = $this->core->getAccess()->resolveDirPath($dir, htmlspecialchars_decode(urldecode($path)));
 
-        if (!$this->core->getAccess()->canI("path.read", ["dir" => $dir, "path" => $path])) {
+        /*if (!$this->core->getAccess()->canI("path.read", ["dir" => $dir, "path" => $path])) {
             $this->core->getOutput()->showError("You do not have access to this file");
             return false;
-        }
+        }*/
```
[the remark this answers is not in the document]
Comment (jchulton) [reply site/app/controllers/MiscController.php]: Done

File: site/app/views/PDFView.php [hunk 7]
```diff
@@ -38,7 +38,10 @@ public function showPDFEmbedded($params) {
         $localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('pdfjs', 'pdf.worker.min.js'), 'vendor');
         $localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('pdf-annotate.js', 'pdf-annotate.min.js'), 'vendor');
         $localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('pdf', 'PDFAnnotateEmbedded.js'), 'js');
-
+        //$this->core->getOutput()->addVendorJs(FileUtils::joinPaths('jspdf', 'jspdf.min.js'));
+        //$localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('html2canvas', 'html2canvas.js'), 'vendor');
+        //$localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('jspdf', 'jspdf.min.js'), 'vendor');
+        //$localjs[] = $this->core->getOutput()->timestampResource(FileUtils::joinPaths('jspdf', 'jspdf.node.min'), 'vendor');
```
[the remark this answers is not in the document]
Comment (jchulton) [reply site/app/views/PDFView.php]: Done

File: site/app/views/PDFView.php [hunk 8]
```diff
@@ -59,4 +62,63 @@ public function showPDFEmbedded($params) {
...
         ]);
     }
+    
+    /**
+     * adds to our buffer a twig output of either student view or grader view.
+     *
+     * @param $gradeable_id
+     * @param $user_id
+     * @param $filename
+     * @param $annotation_jsons
```
[the remark this answers is not in the document]
Comment (jchulton) [reply site/app/views/PDFView.php:69]: Done

File: .setup/install_submitty/install_site_dev.sh [hunk 9]
```diff
@@ -216,7 +216,13 @@ if echo "{$result}" | grep -E -q "package(-lock)?.json"; then
     # twig.js
     mkdir ${VENDOR_FOLDER}/twigjs
     cp ${NODE_FOLDER}/twig/twig.min.js ${VENDOR_FOLDER}/twigjs/
-
+    #html2canvas
+    #mkdir ${VENDOR_FOLDER}/html2canvas
+    #cp ${NODE_FOLDER}/html2canvas/dist/html2canvas.js ${VENDOR_FOLDER}/html2canvas.js
+    #jspdf
+    mkdir ${VENDOR_FOLDER}/jspdf
```
[the remark this answers is not in the document]
Comment (jchulton) [reply .setup/install_submitty/install_site_dev.sh]: Removed the mkdir and cps

**Pull request merged**